incubator/hnc/pkg/reconcilers/object.go

```diff
@@ -96,17 +100,79 @@ func (r *ObjectReconciler) GetGVK() schema.GroupVersionKind {
 	return r.GVK
 }
 
+// SetMode sets the Mode field of an object reconciler and syncs objects in the cluster if needed.
+// The method will return an error if syncs fail.
+func (r *ObjectReconciler) SetMode(ctx context.Context, mode api.SynchronizationMode, log logr.Logger) error {
+	log = log.WithValues("gvk", r.GVK)
+	newMode := r.getValidateMode(mode, log)
+	oldMode := r.Mode
+	if newMode == oldMode {
+		return nil
+	}
+	r.Log.V(1).Info("Changing Mode of the object reconciler", "old", oldMode, "new", newMode)
+	r.Mode = newMode
+	// Currently, there are only two modes: 'ignore' and 'propagate'. If we change from "ignore" mode to 'propagate',
+	// we need to update objects in the cluster.
+	// TODO: We might need to revisit when we want to call SyncCluster if we support more modes in future.
+	if oldMode == api.Ignore {
+		err := r.syncCluster(ctx, r.Log)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// getValidateMode returns a valid api.SynchronizationMode based on the given mode. Please
+// see the comments of api.SynchronizationMode for currently supported modes.
+// If mode is not set, it will be api.Propagate by default. Any unrecognized mode is
+// treated as api.Ignore.
+func (r *ObjectReconciler) getValidateMode(mode api.SynchronizationMode, log logr.Logger) api.SynchronizationMode {
+	switch mode {
+	// TODO: Add api.Remove here. We currently do not support 'remove' mode. A
+	// 'remove' mode will be treated as 'ignore'. We will add api.Remove to the first
+	// case when we start to support the mode.
+	case api.Propagate, api.Ignore:
+		return mode
+	case "":
+		log.V(1).Info("Add mode because mode is not set", "mode", api.Propagate)
+		return api.Propagate
+	default:
+		log.V(1).Info("Update mode because mode is unrecognized", "mode", api.Ignore)
+		return api.Ignore
```

Comment:
These are also probably better as normal logs, not V(1).

They could probably also be made a bit more clear, IMO - for the first one:

`"Unset mode; using "propagate"`

For the second one:

`"Unrecognized mode; using "ignore", "mode", mode`

This ensures that the log includes the _bad_ value that was passed in.

Reply:
Thanks for the explanation. Done.